File: 2024/3/auto.go
//go:build go_auto

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	var solution1, solution2 int
	do := true

	type key struct {
		state int
		char  rune
	}

	t_start := 0
	t_do := 4
	t_dont := 9
	t_mul := 13

	cTr := map[key]int{
		{t_start, 'd'}: 1,
		{1, 'o'}:       2,
		{2, '('}:       3,
		{3, ')'}:       t_do,
		{2, 'n'}:       5,
		{5, '\''}:      6,
		{6, 't'}:       7,
		{7, '('}:       8,
		{8, ')'}:       t_dont,
		{t_start, 'm'}: 10,
		{10, 'u'}:      11,
		{11, 'l'}:      12,
		{12, '('}:      t_mul,
	}

	t_mul1 := 14
	t_mul1c := 15
	t_mul2 := 16
	t_mul2c := 17

	for _, r := range "0123456789" {
		cTr[key{t_mul, r}] = t_mul1
		cTr[key{t_mul1, r}] = t_mul1
		cTr[key{t_mul1c, r}] = t_mul2
		cTr[key{t_mul2, r}] = t_mul2
	}

	cTr[key{t_mul1, ','}] = t_mul1c
	cTr[key{t_mul2, ')'}] = t_mul2c

	/* Visual representation of the automaton (* = digit)

	       ┌────┐   m   ┌───┐
	       │ 10 ◄───────┤ 0 ├t_start
	       └────┘       └─┬─┘
	          │u          │d
	       ┌──▼─┐       ┌─▼─┐
	       │ 11 │       │ 1 │
	       └──┬─┘       └─┬─┘
	          │l          │o
	       ┌──▼─┐       ┌─▼─┐ n   ┌───┐
	       │ 12 │       │ 2 ├─────► 5 │
	       └──┬─┘       └─┬─┘     └─┬─┘
	          │(          │(        │'
	       ┌──▼─┐       ┌─▼─┐     ┌─▼─┐
	       │ 13 ├t_mul  │ 3 │     │ 6 │
	       └──┬─┘       └─┬─┘     └─┬─┘
	          │*          │)        │t
	    ┌──┬──▼─┐       ┌─▼─┐     ┌─▼─┐
	   *└─►│ 14 ├t_mul1 │ 4 ├t_do │ 7 │
	       └──┬─┘       └───┘     └─┬─┘
	          │,                    │(
	       ┌──▼─┐                 ┌─▼─┐
	       │ 15 ├t_mul1c          │ 8 │
	       └──┬─┘                 └─┬─┘
	          │*                    │)
	    ┌──┬──▼─┐                 ┌─▼─┐
	   *└─►│ 16 ├t_mul2           │ 9 ├t_dont
	       └──┬─┘                 └───┘
	          │)
	       ┌──▼─┐
	       │ 17 ├t_mul2c
	       └────┘
	*/

	s, mul, cur, s_pos := 0, 0, 0, 0

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		for i, c := range line {
			if ns, ok := cTr[key{s, c}]; ok {
				s = ns
			} else if ns, ok := cTr[key{0, c}]; ok {
				s = ns
			} else {
				s = 0
			}

			switch s {
			case t_do:
				do = true
			case t_dont:
				do = false
			case t_mul:
				s_pos = i
			case t_mul1c:
				mul, _ = strconv.Atoi(line[s_pos+1 : i])
				s_pos = i
			case t_mul2c:
				cur, _ = strconv.Atoi(line[s_pos+1 : i])
				solution1 += mul * cur
				if do {
					solution2 += mul * cur
				}
			}
		}
	}

	fmt.Println(solution1)
	fmt.Println(solution2)
}

File: 2024/3/auto.py
import sys
import pprint
import collections


def generate():
    def add_word(transitions, word):
        for i, c in enumerate(word):
            transitions[word[:i]][c] = word[: i + 1]

    def compile(transitions):
        tmap = dict()
        for t in transitions:
            if t not in tmap:
                tmap[t] = len(tmap)
            for c in transitions[t]:
                if transitions[t][c] not in tmap:
                    tmap[transitions[t][c]] = len(tmap)
        comp = dict()
        for t in transitions:
            for c in transitions[t]:
                comp[(tmap[t], c)] = tmap[transitions[t][c]]
        return comp, tmap

    s_tr = collections.defaultdict(dict)
    add_word(s_tr, "do()")
    add_word(s_tr, "don't()")
    add_word(s_tr, "mul(")
    for i in range(10):
        cur_str = str(i)
        s_tr["mul("][cur_str] = "mul(*"
        s_tr["mul(*"][cur_str] = "mul(*"
        s_tr["mul(*,"][cur_str] = "mul(*,*"
        s_tr["mul(*,*"][cur_str] = "mul(*,*"
    s_tr["mul(*"][","] = "mul(*,"
    s_tr["mul(*,*"][")"] = "mul(*,*)"

    c_tr, t_map = compile(s_tr)

    pprint.pprint(c_tr)
    pprint.pprint(t_map)


if __name__ == "__main__":
    solution1 = 0
    solution2 = 0
    do = True

    # State constants
    class t:
        start = 0
        do = 4
        dont = 9
        mul = 13
        mul1 = 14
        mul1f = 15
        mul2 = 16
        mul2f = 17

    # Transition map
    cTr = {
        (t.start, "d"): 1,
        (1, "o"): 2,
        (2, "("): 3,
        (3, ")"): t.do,
        (2, "n"): 5,
        (5, "'"): 6,
        (6, "t"): 7,
        (7, "("): 8,
        (8, ")"): t.dont,
        (t.start, "m"): 10,
        (10, "u"): 11,
        (11, "l"): 12,
        (12, "("): t.mul,
    }

    # Add transitions for digits
    for r in "0123456789":
        cTr[(t.mul, r)] = t.mul1
        cTr[(t.mul1, r)] = t.mul1
        cTr[(t.mul1f, r)] = t.mul2
        cTr[(t.mul2, r)] = t.mul2

    cTr[(t.mul1, ",")] = t.mul1f
    cTr[(t.mul2, ")")] = t.mul2f

    state = 0
    mul = 0
    cur = 0
    s_pos = 0

    for line in sys.stdin:
        line = line.rstrip("\n")
        for i, c in enumerate(line):
            if (state, c) in cTr:
                state = cTr[(state, c)]
            elif (0, c) in cTr:
                state = cTr[(0, c)]
            else:
                state = 0

            if state == t.do:
                do = True
            elif state == t.dont:
                do = False
            elif state == t.mul:
                s_pos = i
            elif state == t.mul1f:
                mul_str = line[s_pos + 1 : i]
                mul = int(mul_str)
                s_pos = i
            elif state == t.mul2f:
                cur_str = line[s_pos + 1 : i]
                cur = int(cur_str)
                solution1 += mul * cur
                if do:
                    solution2 += mul * cur

    print(solution1)
    print(solution2)
